Restore ability of Switch to pass through named subconstructs.

hardkrash opened the pull request:
These following constructs were bad:

```python
assert Struct(Switch(lambda ctx: 5, {1:'liz' / Byte, 5:'tom' / Int16ub})).parse(b"\x00\x02") == Container(tom=2)
assert Struct(Switch(lambda ctx: 5, {1:'liz' / Byte, 5:Embedded(Struct('tom' / Int16ub))})).parse(b"\x00\x02") == Container(tom=2)
```

They both returned an empty Container.

Please see tests and commit comments for further details.

hardkrash commented:
s/ass/pass/

Time to go to bed...

hardkrash commented:
I realized that `_parse` of Struct can be simplified by calling each sub construct's `_parse` storing to `subobj` and then inspecting for `flagembedded` or `name` on `sc`.

I'll make the changes tonight (UTC-8).

hardkrash commented:
Ok, Now this is cleaner, and a bit more DRY.

Also, small addition of a __dir__ function crept in...

hardkrash commented:
This might be related to #290.

arekbulski commented:
Honestely, I have no idea what these do. I welcome extentions of the test suite a lot, but I need explanations about the rest of it.

domenzain commented:
@arekbulski , `Switch` has been broken for a while now with many unexpected behaviours around embedding and renaming the switched constructs and the switch itself.

If you look at the test suite modifications, some of them should pass but currently do not:

```python
# This gives an empty Container.
assert Struct(Switch(lambda ctx: 5, {1:'liz' / Byte, 5:Embedded(Struct('tom' / Byte, 'bob' / Byte))})).parse(b"\x00\x02") == Container(tom=0)(bob=2)
# This fails while building with a KeyError
assert Struct(Switch(lambda ctx: 5, {1:'liz' / Byte, 5:Embedded(Struct('tom' / Byte, 'bob' / Byte))})).build(Container(tom=0)(bob=2)) == b"\x00\x02"
# This too gives an empty Container.
assert Struct(Switch(lambda ctx: 5, {1:'liz' / Byte, 5:'tom' / Int16ub})).parse(b"\x00\x02") == Container(tom=2)
# And this fails while building with a KeyError
assert Struct(Switch(lambda ctx: 5, {1:'liz' / Byte, 5:'tom' / Int16ub})).build(Container(tom=2)) == b"\x00\x02"
```

@hardkrash also comments on this case:
```python
Struct('lex' / Switch(lambda ctx: 5, {1:'liz' / Byte, 5:Embedded(Struct('tom' / Byte, 'bob' / Byte))}))
```
Which I think should simply ignore the `'lex'` and keep the individual names of the embedded `Struct` (`'tom'`. `'bob'`) but also rename `'liz'` into `'lex'`.

That is, the rename only makes sense when it is a single construct being renamed and in that case it should work. But otherwise it should not.

arekbulski commented:
I think Switch was never meant to work with Embedded, cant understand how these two would work together anyway. But last two cases are obviously valid. I will look into these.

domenzain commented:
This PR takes care of both.

I would expect `Switch`ing an `Embedded` `Struct` to work like `Embedded`ing a `Switch`ed `Struct`. There is no harm in having the two be commutative. And refusing to take `Embedded`s in a `Switch` is a mild annoyance...

arekbulski commented:
That makes no sense. Switch should at most process Embedded Struct like a normal Struct. And I think that is what it does right now.

domenzain commented:
That's what I meant. A Switch should simply choose the construct that gets parsed without any extra sauce.

Currently that fails:

- when the `Struct` happens to have `flagembedded`. You can confirm this by running the first two tests. They give empty `Container`s.
- when the constructs have names.

arekbulski commented:
I am going to reject this PR. You mixed several unrelated changes which makes this entire thing unreadable. Please... I appreciate the effort, but I dont have an entire day to dissect this. If you find something, make one change and push. 

You did not find a bug, you got the test wrong:
```py3
# name in wrong place
assert Struct(Switch(1, {1:'field'/Byte})).parse(b"\x02") == Container(field=2)
# OK
assert Struct('field'/Switch(1, {1:Byte})).parse(b"\x02") == Container(field=2)
```

domenzain commented:
Not my pull request, nor my commits.

But both those asserts should clearly work... Having to absolutely put the name outside is unintuitive enough to warrant documentation.

hardkrash commented:
It's also a regression from 2.5 and prevents switches from handling contextually aware sub constructs.  I've been traveling and busy the last few weeks, I'll respond with a unsimplified test case that is also broken, but the test you claimed I got wrong is only an oversimplification of the whole issue.  Maybe the more complex case will show you my prospective on the issue.

E.g.  If a flag is set elsewhere in a construct how to select the proper sub structure, without lose of information.

> On Mar 27, 2017, at 15:23, domenzain <notifications@github.com> wrote:
> 
> Not my pull request, nor my commits.
> 
> But both those asserts should clearly work... Having to absolutely put the name outside is unintuitive enough to warrant documentation.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

arekbulski commented:
I know what you are getting at but feel like this is some shenaningan behaviour that somehow worked previously and now is gone. 

The design is simple: Struct needs to have its subcon named, if its anonymous then it does not get put into resulting dict. Using names from subcons was neither intended or documented (by me).

However I reconsidered, and it would seem reasonable to have huge named Structs in separate variables, and simple Switches using them. For example:
```py3
st = 'largedatastructure' / Struct(......)
assert Struct(Switch(1, {1:st})).parse(...) == Container(largedatastructure=...)
```

arekbulski commented:
But this is undoable. Struct only sees an unnamed Switch, and the Switch returns an object and does not signal which of its subcons were used. Outer struct does not have any insight/acces into Switches subcons.

arekbulski commented:
I will still look at some new test cases, @hardkrash .

hardkrash commented:
I realize you don't have too much time to work on this code either, heaven knows I barely have time!
It took me over a month to get time to get back to you.
Here is a more detailed use case.

I have USB hardware data defined by Intel et. al. that returns a bitstruct that is dependent on a mode setting earlier in the bitstruct.  I have slightly simplified this as I am typing form memory, as I can't copy paste the full code.  Variables and names mostly made up.

```python
sc_A = EmbeddedBitStruct(
    'flag_1' / Flag,
    'flag_2' / Flag,
    Padding(8),  # For the sake of brevity, there is one or two enums here.
    'value_2' / BitsInteger(10),
    'param_a' / BitsInteger(10),
)

sc_B = EmbeddedBitStruct(
    'value_1' / BitsInteger(10),
    'value_2' / BitsInteger(10),
    'param_b' / BitsInteger(10),
)

'my_hardware_reg' / BitStruct(
    'mode' / Enum(BitsInteger(2),
        mode_1=1,
        mode_2=2,
        mode_3=3,
    switch(this.mode,
        'mode_1'=sc_A,
        'mode_2'=sc_A,
        'mode_3'=sc_B,
)
```

Additionally this is only one of many examples in my construct 2.5 code that I had update to make this type of bitfield specific data structures work.  These types of modal bitfields and values are popular in micro-controller communication protocols.

As for undoable, it works in the merge request I put together, with test cases.
See commit 324a756 for coping with an embedded struct, then i realized i was too complicated with that commit and added 0a132b0 simplifying my logic.
Are you asking that I merge those two commits?
The defensive programing update is also required as ```all(sc.flagbuildnone for sc in cases.values())``` was evaluating to true and then raising an exception on ```default.flagbuildnone```  I moved it before the tests because I think it caused the tests to raise an exception, but I'm having troubles recalling now.

I added the test cases separate commit as to allow for cherrypicking and verifying with out adding additional steps and allowing git-bisect to work.

The unrelated adding of a __dir__ function to container should be irrelevant but GitHub added it to my open pull request. :-/

domenzain commented:
I second this. If you look at a standard like PTP, the modal structs/bitstructs come again and again. And standing behind PTP (PictBridge) there's all the big consumer camera manufacturers.

It would be nice to support that, especially when the work is already done in this PR.

domenzain commented:
Since this was closed it has lost visibility.
I think the current behaviour of `Switch` is unexpected and the proposed tests should pass.

In your last post here, @arekbulski you mention it is impossible to accomodate the desired result with the current code-base. However, this pull request by @hardkrash adds (brings back, actually) the functionality while not breaking anything.

It is not about whatever is above the `Switch` knowing which construct got chosen, but rather about simply passing the flags or names along when it makes sense.

Could we get this merged or propose a different path to the desired behaviour?
Thanks a lot for your work!

hardkrash commented:
I have a new branch for this feature based off of 881cc323275cf12e368b02c63d1187dbc58ac7de

I'll clean up my commit messages and push that branch to my GitHub repo tomorrow.

There can be some discussion on implementation as well.  I think I simplified it a bit and made it a bit more robust than this implementation.

arekbulski commented:
Oh please make a new PR strictly about Switch/Embedded.
